pkg/report/decompile.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package report
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type DecompilerFlag string
+
+// Extra flags that control the flow of decompilation.
+const (
+	FlagForceArmThumbMode DecompilerFlag = "force_arm_thumb"
+)
+
+const objdumpCallTimeout = 10 * time.Second
+
+type DecompiledOpcode struct {
+	Offset  int
+	Len     int
+	Command string
+	IsBad   bool
+	// Yes, the field below duplicates information and can be inferred from the other fields,
+	// but at this point it does not seem to be reasonable to parse a string and then build
+	// exactly that string from its parts.
+	FullDescription string
+}
+
+// Decompiles a byte array with opcodes into human-readable descriptions.
+// Target must specify the environment from which the opcodes were taken.
+func DecompileOpcodes(raw []byte, flags []DecompilerFlag, target *targets.Target) ([]DecompiledOpcode, error) {
+	args, err := objdumpBuildArgs(flags, target)
+	if err != nil {
+		return nil, err
+	}
+
+	outLines, err := objdumpExecutor(raw, args, target)
+	if err != nil {
+		return nil, err
+	}
+
+	list := objdumpParseOutput(outLines)
+	if len(list) == 0 && len(raw) > 0 {
+		return nil, fmt.Errorf("no instructions found while the total size is %v bytes", len(raw))
+	}
+	return list, nil
+}
+
+func objdumpExecutor(raw []byte, args []string, target *targets.Target) ([]string, error) {
+	file, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp file: %v", err)
+	}
+	defer os.Remove(file.Name())
+	defer file.Close()
+
+	_, err = file.Write(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write to temp file: %v", err)
+	}
+
+	args = append(args, file.Name())
+	cmd := osutil.Command(target.Objdump, args...)
+
+	respBytes, err := osutil.Run(objdumpCallTimeout, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	return strings.Split(string(respBytes), "\n"), nil
```

Comment:
It feels that this strings.Split belongs to objdumpParseOutput b/c it's part of parsing, and for this function it's more natural to return command output as is. It will also remove strings.Split from the test and will allow objdumpParseOutput to use other parsing methods (e.g. bufio.NewScanner is a common way to parse something consisting from multiple lines).